### ISSUE TITLE
Tsyganenko 89,96,01ST,05

### ISSUE THREAD

**Issue by kmrs-iitd** [spacepy/spacepy#787]
I am trying to calculate magnetic field using the data from: https://drive.google.com/file/d/1TXL7Cu--ear8peVNb0yGPE0oeJGugReL/view?usp=sharing ::    this data file is shared through google drive.

pol2=pd.read_csv('~/Documents/T89.csv')

The dataset covers one year, containing 525,600 data points averaged over 1-minute intervals. I’m encountering two main issues:

1. The model can only handle 100,000 data points at a time, so I’ve had to split the dataset into parts (d1, d2, etc.). Is there any way to run the model on the entire dataset in one go?

2. I’ve attached a figure comparing the output from different models. Surprisingly, I found that the magnetic field values were the same across all models, with no dependence on solar wind pressure or Dst. While this could be expected for the T89 model, which depends on Kp instead of pressure or Dst, the other models (T96, T01, T05) should show some sensitivity to solar wind parameters. Below is the code I’m using. The models provide output in GEO coordinates, which I converted to GSM.

### Minimal example to reproduce issue:

```
import spacepy.time as spt
import spacepy.coordinates as spc
import spacepy.irbempy as ib
import pandas as pd
pol2['Date'] = pd.to_datetime(pol2['Date'], errors='coerce')
pol2 = pol2.dropna(subset=['Date'])
In my dataset, the positions are given in GSM coordinates: XGSM, YGSM, and ZGSM.

d1 = pol2[0:100000]
d2 = pol2[100000:200000]
d3 = pol2[200000:300000]
d4 = pol2[300000:400000]
d5 = pol2[400000:500000]
d6 = pol2[500000:]

d_f = pd.DataFrame()

for d_chunk in [d1, d2, d3, d4, d5, d6]:

    t = spt.Ticktock(d_chunk['Date'].tolist(), 'ISO')
    y = spc.Coords(d_chunk[['XGSM', 'YGSM', 'ZGSM']].values.tolist(), 'GSM', 'car', use_irbem=True)

# Used models [ ‘T89’,  ‘T96’,  ‘T01STORM’, ‘T05’]
    
    B_field = ib.get_Bfield(t, y, 'T05')
    
    a1 = pd.DataFrame({
        'Bx': B_field['Bvec'][:,0],
        'By': B_field['Bvec'][:,1],
        'Bz': B_field['Bvec'][:,2],
        'Date': d_chunk['Date'].values
    })
    
    d_f = pd.concat([d_f, a1], ignore_index=True)

d_f = d_f.sort_values(by='Date').reset_index(drop=True)

```

import platform
import sys
import numpy
import scipy
import matplotlib

print(platform.platform())
print(sys.version_info)
print('numpy={0}'.format(numpy.__version__))
print('scipy={0}'.format(scipy.__version__))
print('matplotlib={0}'.format(matplotlib.__version__))

macOS-10.16-x86_64-i386-64bit
sys.version_info(major=3, minor=8, micro=19, releaselevel='final', serial=0)
numpy=1.24.4
scipy=1.10.1
matplotlib=3.7.5

spacepy
Version: 0.7.0a0
downloaded from: pip install git+https://github.com/spacepy/spacepy

Figure:
Panel 1: dBz_obs (Bz_Observed-Bz_IGRF) data from Polar Satellite for a storm on Oct 10-12, 1997.
Panel 2: BzGSM from T89 
Panel 3: BzGSM from T96 
Panel 4: BzGSM from T01 
Panel 5: BzGSM from T05 
Panel 6: Dst from OMNI 
Panel 7: Pres from OMNI 
Panel 8: ZGSM from Polar Satellite 

![Fig](https://github.com/user-attachments/assets/232ab4fe-7b8c-45f6-886f-ee2cae9fd8b8)

**Comment by jtniehof** [comment]
Thanks for the full example.

Here's your data on my machine with T05 in blue and T89 in orange:
![Figure_1](https://github.com/user-attachments/assets/80fc035a-8802-4050-8647-aa45176f0c25)

You have plots of Dst and Pdyn "from OMNI". Is that using the spacepy.omni module or downloading separately? The inputs for the magnetic field models are taken from the omni module. If you import spacepy.omni, do you get the warning "Qin-Denton/OMNI2 data not found in current format."? If so, run toolbox.update to download the latest. See #731, although that shouldn't affect you for the Polar era.

Following on from your example:
```python
o = spacepy.omni.get_omni(pol2['Date'].tolist())
plt.plot(o['UTC'], o['Pdyn'])
```
will let you see the dynamic pressure used in calculating the external B model and compare to what you're plotting (assuming you're not using the omni module for your Pres panel).

I am somewhat nervous about passing the Pandas time type into Ticktock. It seems to work, probably because we get lucky and it can be treated like a string. But `to_datetime` despite the name doesn't make a datetime. If the omni update doesn't get to the bottom of it we can try `d_chunk['Date'].dt.to_pydatetime()`.

As far as the 100000 limit, see #371.

**Comment by kmrs-iitd** [comment]
Thank you very much for your response. 

The Dst and Pdyn plots were not generated using data from the spacepy.omni module; I downloaded that data from CDAWeb/OMNI. 

When importing spacepy.omni, I did not get any warnings. 

In Ticktock I am using now d_chunk['Date'].dt.to_pydatetime().

The file I attached contains Pres, Dst, IMF By/Bz, and Kp. However, when calculating external fields using T89, T96, T01, and T05, SpacePy does not use the solar wind data from my file. Instead, it downloads the data from spacepy.omni module based on the 'Date' column as you mentioned in your comments. 

Is there a way to use Dst, Pdyn, IMF By/Bz, and Kp from my own file? I’d like to compare the external magnetic field from T89 for fixed Kp values, such as B_T89 for Kp = 1 vs. B_T89 for Kp = 5, 6, or 7. 

Similarly, I want to compare B_T96, B_T01, and B_T05 at Dst = -45 with B_T96, B_T01, and B_T05 at Dst = -60, -80, -100, or -120. 

This would enable me to assess how strongly the external magnetic field depends on these parameters.

**Comment by jtniehof** [comment]
Yes, you can directly specify the solar wind / geomagnetic conditions with the `omni` kwarg of [get_Bfield](https://spacepy.github.io/autosummary/spacepy.irbempy.get_Bfield.html#spacepy.irbempy.get_Bfield). [get_Lstar](https://spacepy.github.io/autosummary/spacepy.irbempy.get_Lstar.html#spacepy.irbempy.get_Lstar) has fuller documentation on those keys and what keys are used by what model (#789)--unfortunately all the keys have to be present even if not used (#788).

Going to T01 or T05 requires the W and/or G parameters, which are not simple instantaneous functions of the solar wind or indices.

```python
#!/usr/bin/env python3

import collections

import spacepy.coordinates
import spacepy.irbempy
import spacepy.time


tt = spacepy.time.Ticktock(['2020-01-01T00:00'] * 3)
loci = spacepy.coordinates.Coords([[6, 0, 0]] * 3, 'GSE', 'car')

omnivals = collections.defaultdict(
    lambda: [0.] * 3,
    Kp=[0., 4., 9.])
print(spacepy.irbempy.get_Bfield(tt, loci, 'T89', omnivals=omnivals))

omnivals = collections.defaultdict(
    lambda: [0.] * 3,
    Dst=[-90, -20, 5],
    Pdyn=[2.] * 3,
    ByIMF=[0.] * 3,
    BzIMF=[0.] * 3,
)
print(spacepy.irbempy.get_Bfield(tt, loci, 'T96', omnivals=omnivals))
```

**Comment by jtniehof** [comment]
The "omni" values in the SpacePy omni module are not the values directly from OMNIWeb. They're the processed values according to [Qin et al](https://doi.org/10.1029/2006SW000296). We're working on getting this hosted with SPDF (#731)--previously they were hosted by ViRBO.

**Comment by kmrs-iitd** [comment]
Thank you very much, now I can use directly the solar wind / geomagnetic conditions. In all the examples (Bfield, Lstar, Bmirror) provided in the irbempy documentation, the GEO coordinate system is used and the output is in GEO. Can I use X, Y, Z in GSM/GSE/SM coordinates for the Bfield (or other) and obtain Bvec in GSM/GSE/SM?

Thanks

**Comment by jtniehof** [comment]
get_Bfield [always returns GEO](https://spacepy.github.io/autosummary/spacepy.irbempy.get_Bfield.html#spacepy.irbempy.get_Bfield) (#676) although the inputs can be in any coordinates. #377 has a workaround until we implement returning the same vector.

**Comment by kmrs-iitd** [comment]
I sincerely appreciate your support.
